### PR TITLE
Implement Spike retriever

### PR DESCRIPTION
**Pull request by samuelgarcia**
Need #1941 to be merged before.

SpikeRetriever class is usefull to inject a sorting object in the node pipepline mechanisim.
It allows to compute some post processing with the same machinery used for sorting components.
This is a first step to totaly refactor: 
  * compute_spike_locations()
  * compute_amplitude_scalings()
  * compute_spike_amplitudes()
  * compute_principal_components()

### CODE REVIEW

**Comment by samuelgarcia** [comment]
@alejoe91 this is ready to be reviewed/merged

File: src/spikeinterface/core/node_pipeline.py [hunk 1]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+      * compute_spike_locations()
+      * compute_amplitude_scalings()
+      * compute_spike_amplitudes()
+      * compute_principal_components()
+
+
+    recording:
+
+    sorting:
+
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:154]: missing

File: src/spikeinterface/core/node_pipeline.py [hunk 2]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
 
 # this is not implemented yet this will be done in separted PR
 class SpikeRetriever(PeakSource):
-    pass
+    """
+    This class is usefull to inject a sorting object in the node pipepline mechanisim.
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:142]:
```suggestion
    This class is useful to inject a sorting object in the node pipepline mechanism.
```

File: src/spikeinterface/core/node_pipeline.py [hunk 3]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+    It allows to compute some post processing with the same machinery used for sorting components.
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:143]:
```suggestion
    It allows to compute some post-processing steps with the same machinery used for sorting components.
```

File: src/spikeinterface/core/node_pipeline.py [hunk 4]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+    This is a first step to totaly refactor:
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:144]: Docstring should be as static as possible...
```suggestion
    This is used by:
```

File: src/spikeinterface/core/node_pipeline.py [hunk 5]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+      * compute_spike_amplitudes()
+      * compute_principal_components()
+
+
+    recording:
+
+    sorting:
+
+    channel_from_template: bool (default True)
+        If True then the channel_index is infered from template and extremum_channel_inds must be provided.
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:156]:
```suggestion
        If True, then the channel_index is inferred from the template and `extremum_channel_inds` must be provided.
```

File: src/spikeinterface/core/node_pipeline.py [hunk 6]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+        If False every spikes compute its own channel index given a radius around the template max channel.
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:157]:
```suggestion
        If False, the max channel is computed for each spike given a radius around the template max channel.
```

File: src/spikeinterface/core/node_pipeline.py [hunk 7]
```diff
@@ -138,7 +138,100 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+                    local_peaks[i]["channel_index"] = chans[np.argmax(sparse_wfs)]
+                elif self.peak_sign == "both":
+                    local_peaks[i]["channel_index"] = chans[np.argmax(np.abs(sparse_wfs))]
+
+        # TODO: "amplitude" ???
+
+        return (local_peaks,)
+
+
+def sorting_to_peak(sorting, extremum_channel_inds):
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:226]:
```suggestion
def sorting_to_peaks(sorting, extremum_channel_inds):
```

File: src/spikeinterface/core/tests/test_node_pipeline.py [hunk 8]
```diff
@@ -12,9 +12,10 @@
 from spikeinterface.core.node_pipeline import (
     run_node_pipeline,
     PeakRetriever,
+    SpikeRetriever,
     PipelineNode,
     ExtractDenseWaveforms,
-    base_peak_dtype,
+    sorting_to_peak,
```
Comment (alejoe91) [inline src/spikeinterface/core/tests/test_node_pipeline.py:18]:
```suggestion
    sorting_to_peaks,
```

File: src/spikeinterface/core/tests/test_node_pipeline.py [hunk 9]
```diff
@@ -71,106 +72,115 @@ def compute(self, traces, peaks, waveforms):
 def test_run_node_pipeline():
     recording, sorting = generate_ground_truth_recording(num_channels=10, num_units=10, durations=[10.0])
 
-    job_kwargs = dict(chunk_duration="0.5s", n_jobs=2, progress_bar=False)
+    # job_kwargs = dict(chunk_duration="0.5s", n_jobs=2, progress_bar=False)
+    job_kwargs = dict(chunk_duration="0.5s", n_jobs=1, progress_bar=False)
```
Comment (alejoe91) [inline src/spikeinterface/core/tests/test_node_pipeline.py:76]: why not 2 jobs?

File: src/spikeinterface/core/tests/test_node_pipeline.py [hunk 10]
```diff
@@ -71,106 +72,115 @@ def compute(self, traces, peaks, waveforms):
...
-    assert np.array_equal(denoised_waveforms_rms2, denoised_waveforms_rms3)
-
-    # Test pickle mechanism
-    for node in nodes:
-        import pickle
-
-        pickled_node = pickle.dumps(node)
-        unpickled_node = pickle.loads(pickled_node)
+
+    # test with 2 diffrents first node
```
Comment (alejoe91) [inline src/spikeinterface/core/tests/test_node_pipeline.py:100]:
```suggestion
    # test with 3 differents first nodes
```

**Review by alejoe91 — CHANGES_REQUESTED** [review]
Thanks @samuelgarcia 

A few comments and suggestions!

File: src/spikeinterface/core/node_pipeline.py [hunk 11]
```diff
@@ -138,7 +138,99 @@ def compute(self, traces, start_frame, end_frame, segment_index, max_margin):
...
+    """
+
+    def __init__(
+        self, recording, sorting, channel_from_template=True, extremum_channel_inds=None, radius_um=50, peak_sign="neg"
+    ):
+        PipelineNode.__init__(self, recording, return_output=False)
+
+        self.channel_from_template = channel_from_template
+
+        assert extremum_channel_inds is not None, "SpikeRetriever need the dict extremum_channel_inds"
```
Comment (alejoe91) [inline src/spikeinterface/core/node_pipeline.py:174]:
```suggestion
        assert extremum_channel_inds is not None, "SpikeRetriever needs the extremum_channel_inds dictionary"
```

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**